### PR TITLE
Code Review Skills

### PR DESCRIPTION
**Pull request by swapnil-nagar**
# Code Review Skills

## Summary

Adds three repository-scoped GitHub Copilot skills for the Azure Functions Node.js worker. These skills provide domain-aware validation, security analysis, and mandatory human-review gates without changing worker runtime behavior.

### Host and Worker Domain Skill

- Documents Host-worker architecture and lifecycle.
- Covers gRPC, invocation, indexing, module loading, hooks, and Node.js process behavior.
- Maps affected components to focused tests and broader validation.
- Protects protobuf generation and compatibility invariants.

### Security Skill

- Defines repository-specific trust boundaries and protected assets.
- Traces potentially untrusted data from source to sensitive sink.
- Identifies credential disclosure, injection, isolation, availability, and supply-chain risks.
- Requires evidence-backed findings and regression tests.
- Includes safeguards against expected-behavior false positives.

### Critical-Path Human Review Skill

- Identifies changes requiring qualified human intervention.
- Defines Gate A for approval before implementation.
- Defines Gate B for review before acceptance.
- Produces a structured `HUMAN REVIEW REQUIRED` packet.
- Prevents Copilot from treating tests as human approval.
- Integrates with the repository's CODEOWNERS policy.

## Usage

The skills are automatically discovered from `.github/skills/` and can also be invoked manually:

- `/azure-functions-node-worker`
- `/azure-functions-node-worker-security`
- `/azure-functions-node-worker-critical-path`

## Validation

- Skill frontmatter and discovery metadata validated.
- Relative reference links verified.
- Markdown diagnostics and whitespace checks passed.
- No runtime tests were required because only Copilot customization files were added.

### CODE REVIEW

**Review by hallvictoria — APPROVED** [review]

**Comment by TsuyoshiUshio** [comment]
My first comment is add `AGENTS.md` for make sure it invoked or add `Prompts(slash command)` If you want to use it only when you want to use it. :) 

But if you are not familiar with it I'd happy to add it with follow up PRs.

File: .github/skills/azure-functions-node-worker-critical-path/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+argument-hint: '[diff, file, proposed change, or PR]'
+user-invocable: true
+disable-model-invocation: false
+---
+
+# Azure Functions Node.js Worker Critical Path
+
+This skill is a review gate, not a general code review. Its purpose is to identify changes for which automated reasoning and tests are insufficient and a qualified human must make or approve the decision.
+
+Apply `azure-functions-node-worker` for domain validation. Apply `azure-functions-node-worker-security` whenever the change affects a trust boundary, executable code selection, credentials, process isolation, availability controls, or dependencies.
```
Comment (TsuyoshiUshio) [inline .github/skills/azure-functions-node-worker-critical-path/SKILL.md:13]: No need to fix it now, However, this section has dependency for the other skills. If you expect to invoke these skills always, you can make it one skill and move the contents into reference files. 
If you have chance to use the skills individually, you can keep it but better to clarify the how to organize the skills structure.  Skills are progressive disclosure so that metadata are included into the context always. 

so that reducing the number of Skills (that is invoked together) is a good strategy. 

Also naming wise, if it is `review` skill you can make it clear to do `pr-review` or some other stuff. It is clear the purpose and agent can invoke when you want to review or end of the development gate. 

But it is not blocker. :)

File: .github/skills/azure-functions-node-worker-critical-path/references/critical-paths.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+- refactoring dispatch or a handler while preserving message shapes
+- fixing retries or ESM/CommonJS loading without expanding valid targets
+- changing converters while preserving documented binding semantics
+- changing hook execution implementation without changing ordering or error policy
+- optimizing hot invocation code, allocations, or blocked-event-loop monitoring
+- extending secret patterns without reducing existing coverage
+- adding validation that enforces an already documented Host invariant
+- changing tests or fixtures that encode a critical contract
+
+Gate B requires a focused regression plus broad repository validation. Security-sensitive cases also require the security skill's source-to-sink analysis.
```
Comment (TsuyoshiUshio) [inline .github/skills/azure-functions-node-worker-critical-path/references/critical-paths.md:45]: How about let agent check the upstream critical library change?  For example, when we change the grpc, let the agent see the diff or change logs upstream library and if some important change happens, let us know or test. 

```
Dependency upgrade requirements

For upgrades affecting critical-path dependencies:

- Review upstream release notes and changelog.
- Identify breaking changes, behavior changes, and security fixes.
- Document which upstream changes may affect the Node.js worker.
- Map each identified behavior change to validation evidence or regression tests.
- Record the source version and target version reviewed.
- If upstream documentation is unavailable, mark the dependency analysis incomplete.
```

Comment (TsuyoshiUshio) [reply .github/skills/azure-functions-node-worker-critical-path/references/critical-paths.md:45]: This is also not blocker.

File: .github/skills/azure-functions-node-worker-critical-path/references/critical-paths.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+When generated output changes unexpectedly, escalate rather than classify it as harmless.
+
+## Cross-Repository Review
+
+Require evidence or ownership input from related repositories when changing shared behavior:
+
+- `azure-functions-host` for RPC messages, capabilities, startup, and lifecycle contracts
+- `azure-functions-nodejs-library` for programming-model and core API behavior
+- `azure-functions-nodejs-e2e-tests` for ecosystem compatibility coverage
+- `azure-functions-core-tools` for local Host startup and debugging behavior
```
Comment (TsuyoshiUshio) [inline .github/skills/azure-functions-node-worker-critical-path/references/critical-paths.md:65]: `azure-functions-nodejs-extensions` ?

File: .github/skills/azure-functions-node-worker-security/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+Test boundary values appropriate to the flow, including missing/null values, maximum sizes, duplicate or unexpected map keys, malformed encoded data, mixed path separators, absolute/drive-relative/UNC paths, traversal segments, symlinks, circular error objects, secret-bearing URLs, concurrent invocations, specialization during outstanding work, and hooks that throw or never settle.
+
+Do not create exploit files outside the test workspace, access real credentials, contact production endpoints, or print discovered secret values. Use synthetic values and repository test fixtures.
+
+### 5. Validate the Fix or Finding
+
+Run the nearest security regression first, then the domain-required checks. Typical focused commands are:
+
+```powershell
+npx mocha -r ts-node/register "test/errors.test.ts"
```
Comment (TsuyoshiUshio) [inline .github/skills/azure-functions-node-worker-security/SKILL.md:92]: Do we need `npx`? since we have `package.json`? Is there any reason to use it?

File: .github/skills/azure-functions-node-worker/references/architecture.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+## System Boundaries
+
+- Azure Functions Host: starts the worker, owns the RPC protocol, sends lifecycle and invocation messages, and consumes responses and logs.
+- Node.js worker: owns the gRPC client, process lifecycle, app state, dispatch, compatibility behavior, and the bridge to the programming model.
+- `@azure/functions` and `@azure/functions-core`: own the programming model used to register and invoke v4 functions.
+- Function application: supplies package metadata and executable CommonJS or ESM user code.
+- Language-worker protobuf: defines the shared wire contract. Generated JavaScript and TypeScript files are artifacts, not source of truth.
+
+Related repositories include `azure-functions-host`, `azure-functions-nodejs-library`, `azure-functions-nodejs-e2e-tests`, and `azure-functions-core-tools`. A local unit test cannot prove cross-repository compatibility when the shared contract changes.
```
Comment (TsuyoshiUshio) [inline .github/skills/azure-functions-node-worker/references/architecture.md:13]: add `azure-functions-nodejs-extensions`.

File: .github/skills/azure-functions-node-worker/references/architecture.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+1. Confirm the corresponding Host contract and compatibility plan.
+2. Edit `.proto` source only.
+3. Run `npm run gen`.
+4. Inspect every generated diff.
+5. Add handler/converter tests for old and new shapes as compatibility requires.
+6. Run lint, all unit tests, build, and Host/e2e integration where available.
+
+## Focused Tests
+
+| Behavior                           | Primary tests                                                 |
```
Comment (TsuyoshiUshio) [inline .github/skills/azure-functions-node-worker/references/architecture.md:136]: it might be a good idea to move the architecture.md to the top level and 
let agent know through the AGENTS.md

and direct when we change the architecture, let the agent reflect the information like followings. 
We need make it up to date.

File: .github/skills/azure-functions-node-worker/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+- Generated protocol outputs match their `.proto` sources and are reviewed as generated artifacts.
+
+### 4. Implement and Test the Behavior
+
+Keep the change at the owning abstraction. Add or update the nearest test that observes the externally meaningful result, not only an internal call.
+
+Use the test map in [architecture.md](./references/architecture.md). Prefer the smallest executable check first, for example:
+
+```powershell
+npx mocha -r ts-node/register "test/Worker.test.ts"
```
Comment (TsuyoshiUshio) [inline .github/skills/azure-functions-node-worker/SKILL.md:76]: this section uses npx as well. Any reasons?

File: .github/skills/azure-functions-node-worker/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
+---
+name: azure-functions-node-worker
```
Comment (TsuyoshiUshio) [inline .github/skills/azure-functions-node-worker/SKILL.md:2]: Skill should include `verb` elements.  Since skills are invoked by purpose.  So that this skill might be `azure-functions-node-worker-change-validation` or something. (it is ok to make it small) 

I also feel, better to reference `azure-functions-node-worker-critical-path` and `azure-functions-node-worker-security` from this skill. (or make these reference files.)

**Comment by swapnil-nagar** [comment]
Thanks @TsuyoshiUshio. This PR is intentionally an experimental setup so we can evaluate the three skills independently, including relevance-based model invocation and their built-in slash commands. I am keeping the skills separate for this experiment and clarifying when each one should be loaded.

I will not add `AGENTS.md` in this PR. We can explore an always-on `AGENTS.md` and, if useful, a composite review prompt in a follow-up PR after we learn from this setup.

**Review by TsuyoshiUshio — COMMENTED** [review]
I'm ok to merge it in. Since I believe skills are not about perfection but improvement/iteration. If you start using it immediately, I can approve.  Let me know.

My recommendation is following:
* Move `architecture.md` from reference to toplevel file to keep it updated.
* Dependency improvement among the skills. 

That is the only blocker from my point of view. Otherwise it is nit picking. :)

**Review by TsuyoshiUshio — APPROVED** [review]
